// @vuepress/markdown/lib/plugins/codePlugin/languages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.languageYaml = exports.languageVue = exports.languageTypescript = exports.languageStylus = exports.languageRust = exports.languageRuby = exports.languagePython = exports.languageMarkdown = exports.languageKotlin = exports.languageJavascript = exports.languageHtml = exports.languageBash = void 0;
exports.languageBash = {
    name: 'bash',
    ext: 'sh',
    aliases: ['bash', 'sh', 'shell'],
};
exports.languageHtml = {
    name: 'markup',
    ext: 'html',
    aliases: ['html'],
};
exports.languageJavascript = {
    name: 'javascript',
    ext: 'js',
    aliases: ['javascript', 'js'],
};
exports.languageKotlin = {
    name: 'kotlin',
    ext: 'kt',
    aliases: ['kotlin', 'kt'],
};
exports.languageMarkdown = {
    name: 'markdown',
    ext: 'md',
    aliases: ['markdown', 'md'],
};
exports.languagePython = {
    name: 'python',
    ext: 'py',
    aliases: ['py', 'python'],
};
exports.languageRuby = {
    name: 'ruby',
    ext: 'rb',
    aliases: ['rb', 'ruby'],
};
exports.languageRust = {
    name: 'rust',
    ext: 'rs',
    aliases: ['rs', 'rust'],
};
exports.languageStylus = {
    name: 'stylus',
    ext: 'styl',
    aliases: ['styl', 'stylus'],
};
exports.languageTypescript = {
    name: 'typescript',
    ext: 'ts',
    aliases: ['ts', 'typescript'],
};
exports.languageVue = {
    name: 'markup',
    ext: 'vue',
    aliases: ['vue'],
};
exports.languageYaml = {
    name: 'yaml',
    ext: 'yml',
    aliases: ['yaml', 'yml'],
};
